File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int i = n-1;
        while(i > 0){
            if(nums[i-1] < nums[i])
                break;
            i--;
        }
        if(i == 0){
            sort(nums.begin(), nums.end());
            return ;
        }
        int pointValue = nums[i-1];
        int currGreat = INT_MAX;
        int gInd;
        int j = i;
        while(j < n){
            if(nums[j] > pointValue && nums[j] < currGreat){
                currGreat = nums[j];
                gInd = j;
            }
            j++;
        }
        swap(nums[i-1], nums[gInd]);
        sort(nums.begin() + i , nums.end());
    }
};

File: 31-next-permutation/next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int point = -1;

        // Find the pivot
        for (int i = n - 1; i > 0; i--) {
            if (nums[i - 1] < nums[i]) {
                point = i - 1;
                break;
            }
        }
        if (point == -1) {
            Arrays.sort(nums);
            return;
        }
        int j = n - 1;

        while (nums[j] <= nums[point]) {
            j--;
        }

        // Swap
        int temp = nums[point];
        nums[point] = nums[j];
        nums[j] = temp;
        int left = point + 1;
        int right = n - 1;

        while (left < right) {
            temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;

            left++;
            right--;
        }
    }
}
